File: zigzagiterator.py
class ZigzagIterator(object):

    def __init__(self, v1, v2):
        """
        Initialize your data structure here.
        :type v1: List[int]
        :type v2: List[int]
        """
        self.res = []
        self.index = 0
        i = 0
        while i < len(v1) or i < len(v2):
            if i < len(v1):
                self.res.append(v1[i])
            if i < len(v2):
                self.res.append(v2[i])
            i += 1
        self.len = len(self.res)

    def next(self):
        """
        :rtype: int
        """
        return self.res[self.index-1]
        

    def hasNext(self):
        """
        :rtype: bool
        """
        if self.index < self.len:
            self.index += 1
            return True
        return False
        

# Your ZigzagIterator object will be instantiated and called as such:
# i, v = ZigzagIterator(v1, v2), []
# while i.hasNext(): v.append(i.next())

File: zigzagiterator.cpp
class ZigzagIterator {
public:
    ZigzagIterator(vector<int>& v1, vector<int>& v2) {
        if (!v1.empty()) que.push(make_pair(v1.begin(), v1.end()));
        if (!v2.empty()) que.push(make_pair(v2.begin(), v2.end()));
    }

    int next() {
        auto it = que.front().first;
        auto eit = que.front().second;
        que.pop();
        if (it + 1 != eit) que.push(make_pair(it+1, eit));
        return *it;
    }

    bool hasNext() {
        return !que.empty();
    }
private:
    queue<pair<vector<int>::iterator, vector<int>::iterator>> que;
};

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator i(v1, v2);
 * while (i.hasNext()) cout << i.next();
 */
